### ISSUE TITLE
Fallback to closest possible constraint when default for camera cannot be met

### ISSUE THREAD

**Issue by dreamerns** [twilio/video-quickstart-android#68]
Currently, if the default constraints are not met(640x480@30fps) or if the developer provides custom constraints that cannot be met for the camera the call to `LocalVideoTrack.addVideoTrack(...)` returns null.

Rather than returning null we will attempt to choose a configuration that is close to this default and provide a warning in the log specifying which configuration we selected.

**Comment by dreamerns** [comment]
I think we should also fallback to default constraint if custom one can't be met ?

**Comment by erito** [comment]
This is great, we were talking about building this out ourselves by inspecting the devices capabilities while compiling the constraints into the builder.  At the very least it should check to see if its capable of meeting the constraints, and falling back to the default `VideoConstraint` if this isn't possible.

**Comment by idelgado** [comment]
@erito I like that suggestion as well. We'll discuss this internally and provide a more detailed description of our proposal to ensure it solves the problem in the least surprising way for you and other developers.

**Comment by aaalaniz** [comment]
Hey @erito,

We currently have an internal PR where I've attempted to address some of the `VideoConstraints` issues. Would love to hear your thoughts on our approach. Below are the two scenarios I've tried to address.

**No Constraints are Provided**
This is the primarily the most common use case for adding a `LocalVideoTrack`. Previously we would rely on default constraints of 640x480@30 FPS and assumed the `VideoCapturer` supported these constraints. I've implemented a closest supported algorithm based on a `VideoCapturer#getSupportedFormats()`. `LocalMedia` will now find the closest supported format to 640x480@30 FPS to build default constraints.

**Custom Constraints are Provided**
When a developer provides custom constraints we now perform a validation that confirms that the provided constraints are supported by the `VideoCapturer`. If the constraints cannot be met then we fallback to a supported format closest to 640x480@30 FPS.

**Comment by erito** [comment]
That sounds preferable to the current implementation for sure, also sounds like the logic for closest supported Video Capture format gets reused as the fallback for the Custom Constraints implementation, which is great.

One additional enhancement that might be particularly useful is an equals()/hashcode() implementation for `VideoConstraints` to help verify the desired constraints were added.  I know theres a `VideoConstraints` property inside the returned `LocalVideoTrack`, but its hard to inspect all the attributes without having to do something like this:

```java

public void addTrack(VideoConstraints desiredConstraints) {

    LocalVideoTrack track = mLocalMedia.addVideoTrack(!mLocalVideoDisabled, mCameraCapture);
    VideoConstraints acceptedConstraints = track.getVideoConstraints();
    if (desiredConstraints.getAspectRatio() == acceptedConstraints.getAspectRatio() &&
	desiredConstraints.getMaxFps() == acceptedConstraints.getMaxFps() &&
	desiredConstraints.getMinFps() == acceptedConstraints.getMinFps() &&
	desiredConstraints.getMinVideoDimensions() == acceptedConstraints.getMinVideoDimensions() &&
        desiredConstraints.getMaxVideoDimensions() == acceptedConstraints.getMaxVideoDimensions()) {
            Log.d("Constraints", "Our constraints were accepted!");
    }
}

```

**Comment by aaalaniz** [comment]
@erito Thanks for your feedback.

In the current solution if constraints cannot be applied then we fallback to the closest to 640x480@30 FPS and **NOT** the closest to the custom constraints. I started down the path of the closest to custom constraints approach and was worried about the growing logic to accommodate something like 1280x720 - 1920x1080 16:9 for a `VideoCapturer` that has no 16:9 or HD resolutions. However, I want to continue iterating on this approach because I think it might be a better experience for applying constraints.

As part of testing this work we've added `equals(Object o)` and `hashCode()` for to `VideoConstraints`, `VideoDimensions`, and `AspectRatio` 😄
